### PR TITLE
CRM for ECMP and VoQ System

### PR DESCRIPTION
**Pull request by JaiOCP**
This PR brings in CRM query attributes for IP only Nexthop Group, Members and VoQ System.
Old PR is closed because of enum shifts:
https://github.com/opencomputeproject/SAI/pull/1763

Here is a workflow for these removed attributes.
// Available System Ports
sai_object_type_get_availability(<switch_id>, SAI_OBJECT_TYPE_SYSTEM_PORT, 0,
NULL, &count)

// Available Fabric Ports
sai_attribute_t attr[1];
attr[0].id = SAI_PORT_ATTR_TYPE;
attr[0].value = SAI_PORT_TYPE_FABRIC;
sai_object_type_get_availability(<switch_id>, SAI_OBJECT_TYPE_PORT, 1,
attr, &count)

// Avaiable (underlay) IP nexthopgroups
sai_attribute_t attr[1];
attr[0].id = SAI_NEXT_HOP_GROUP_ATTR_TYPE;
attr[0].value = SAI_NEXT_HOP_GROUP_TYPE_DYNAMIC_UNORDERED_ECMP;

attr[1].id = SAI_NEXT_HOP_GROUP_ATTR_HIERARCHICAL_NEXTHOP;
attr[1].value = false;
sai_object_type_get_availability(<switch_id>,
SAI_OBJECT_TYPE_NEXTHOPGROUP, 2, attr, &count)

// Avaiable (overlay) IP+Tunnel nexthopgroups
attr[1].id = SAI_NEXT_HOP_GROUP_ATTR_HIERARCHICAL_NEXTHOP;
attr[1].value = true;
sai_object_type_get_availability(<switch_id>,
SAI_OBJECT_TYPE_NEXTHOPGROUP, 2, attr, &count)

For Systems with dedicated mcast and cast.
sai_attribute_t attr[1];
attr[0].id = SAI_QUEUE_ATTR_TYPE;
attr[0].value = SAI_QUEUE_TYPE_UNICAST_VOQ; // or SAI_QUEUE_TYPE_MULTICAST_VOQ
sai_object_type_get_availability(<switch_id>, SAI_OBJECT_TYPE_QUEUE, 1,
attr, &count)

### CODE REVIEW

File: inc/saiswitch.h [hunk 1]
```diff
@@ -2834,6 +2834,22 @@ typedef enum _sai_switch_attr_t
...
+     */
+    SAI_SWITCH_ATTR_AVAILABLE_IP_NEXT_HOP_GROUP_MEMBER_ENTRY,
+
+    /**
+     * @brief Available number of VOQ in the VOQ System (chassis)
+     *
+     * @type sai_uint32_t
+     * @flags READ_ONLY
+     */
+    SAI_SWITCH_ATTR_AVAILABLE_SYSTEM_VOQS,
```
Comment (jasmeetbagga) [inline inc/saiswitch.h:2851]: Do we still need this attribute? vs just get this info like so
``
sai_attribute_t attr[1];
attr[0].id = SAI_QUEUE_ATTR_TYPE;
attr[0].value = SAI_QUEUE_TYPE_UNICAST_VOQ; // or SAI_QUEUE_TYPE_MULTICAST_VOQ
sai_object_type_get_availability(<switch_id>, SAI_OBJECT_TYPE_QUEUE, 1,
attr, &count)

Comment (JaiOCP) [reply inc/saiswitch.h:2851]: There are two kind of VoQ architecture.
1. Where there is dedicated unicast and multicast voqs. In this case existing attribute SAI_QUEUE_TYPE_UNICAST_VOQ and SAI_QUEUE_TYPE_MULTICAST_VOQ should be queried.
2. Where is a common VoQ pool and dynamically unicast and multicast VOQs are allocated. In this case new attribute SAI_SWITCH_ATTR_AVAILABLE_SYSTEM_VOQS should be queried.

**Review by jasmeetbagga — APPROVED** [review]

File: inc/saiswitch.h [hunk 2]
```diff
@@ -2834,6 +2834,22 @@ typedef enum _sai_switch_attr_t
      */
     SAI_SWITCH_ATTR_SWITCH_ISOLATE,
 
+    /**
+     * @brief Available IP Next hop group member entries
+     *
+     * @type sai_uint32_t
+     * @flags READ_ONLY
+     */
+    SAI_SWITCH_ATTR_AVAILABLE_IP_NEXT_HOP_GROUP_MEMBER_ENTRY,
```
Comment (rlhui) [inline inc/saiswitch.h:2843]: @JaiOCP , can we please address the comment from @rck-innovium  in previous PR https://github.com/opencomputeproject/SAI/pull/1763

pasting the below one here:

We need to understand the CRM types for "IP" nexthop group and nexthop group members. Today in SAI, nexthop group does not really differ based on the constituent nexthop's type.

Comment (rlhui) [reply inc/saiswitch.h:2843]: @JaiOCP , if this is corresponding to next-hop-group-member types, the types are SAI_SWITCH_ATTR_AVAILABLE_NEXT_HOP_GROUP_MEMBER_ENTRY which is not differentiating IP or non-IP.
The IP, tunnel type etc. is for next-hop entry .

Comment (JaiOCP) [reply inc/saiswitch.h:2843]: > @JaiOCP , can we please address the comment from @rck-innovium in previous PR #1763
> 
> pasting the below one here:
> 
> We need to understand the CRM types for "IP" nexthop group and nexthop group members. Today in SAI, nexthop group does not really differ based on the constituent nexthop's type.

NHG differs based on the member types. This change was introduced using SAI_NEXT_HOP_GROUP_ATTR_HIERARCHICAL_NEXTHOP bool attribute. Default is IP+Tunnel for backward compatibility.


SAI_SWITCH_ATTR_AVAILABLE_IP_NEXT_HOP_GROUP_MEMBER_ENTRY is for the IP only NGH members.

Comment (JaiOCP) [reply inc/saiswitch.h:2843]: > @JaiOCP , if this is corresponding to next-hop-group-member types, the types are sai_next_hop_group_type_t which is not differentiating IP or non-IP. The IP, tunnel type etc. is for next-hop entry .

HI Rita,

We have the new attribute introduced by Hierarchical ECMP that depicts what kind of NHG it is. 
WIth HECMP we have NHG consists only IP+Tunnel NH member entries or IP only entries. Defaults are always IP+Tunnel hence the new attribute is added to get the IP only available member entries.

Comment (rlhui) [reply inc/saiswitch.h:2843]: then how to query for available NHG entries that for the type of "IP+Tunnel"?  And also total NHG entries?  Can we also clarify the usage of SAI_SWITCH_ATTR_AVAILABLE_NEXT_HOP_GROUP_MEMBER_ENTRY vs this one?

Comment (JaiOCP) [reply inc/saiswitch.h:2843]: Available NHG is captured in the PR comments.

Available NHG entries would be queried using the existing attribute SAI_SWITCH_ATTR_AVAILABLE_NEXT_HOP_GROUP_MEMBER_ENTRY. This is the existing attribute and is for IP+Tunnel NHG member entries.

New read only attribute  SAI_SWITCH_ATTR_AVAILABLE_IP_NEXT_HOP_GROUP_MEMBER_ENTRY will give the IP only NHG member entries.

Comment (rlhui) [reply inc/saiswitch.h:2843]: there can be a mix of "IP+Tunnel" plus "IP only" NHG entries existing, right?  in this case, can we define how the available "IP+tunnel" NHG entries can be queried?

Comment (prsunny) [reply inc/saiswitch.h:2843]: SAI_SWITCH_ATTR_AVAILABLE_NEXT_HOP_GROUP_MEMBER_ENTRY, just to confirm, will it cover all sorts for NEXT_HOP member types? Sonic queries this currently and relies on this for creating any type of nexthop group member.

Comment (j-bos) [reply inc/saiswitch.h:2843]: Would it be possible to use the generic CRM for this as well? My thought:

- Add a read-only attribute to the nhg member which also has ``@isresource`` to reflect whether its group is L1 or L2.
- Use that attribute via the generic crm apis.
- This way it is possible to query for total, by not passing the attribute, or L1 or L2, by passing a value for the attribute.

**Comment by kcudnik** [comment]
please resolve conflicts

File: doc/CRM-WorkFlow.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+## NextHopGroup and Member Query <a name="introduction2"></a>
+SAI supports default as NHG that contains tunnel and IP nexthop members. In this case usually there is a single HW table for NHG. This HW architecture is defined using SAI_NEXT_HOP_GROUP_ATTR_HIERARCHICAL_NEXTHOP with default value as true.
+
+For cases where HW has two different physical tables, one to host Tunnel+IP overlay nexthops and other to host IP only underlay nexthops; SAI_NEXT_HOP_GROUP_ATTR_HIERARCHICAL_NEXTHOP can be quried using capability query API.
+
+Once NOS has determined the HW architecture it can make following queries to determine the available NHG and NHG member entries.
+
+![](./figures/HECMP.png "Figure 1: Hierarchical ECMP")
+__Figure 1: Hierarchical ECMP__
+
```
Comment (rck-innovium) [inline doc/CRM-WorkFlow.md:62]: @JaiOCP  If there is a 3-rd level of Nexthop group, how will we able to get the available resource count?

**Comment by JaiOCP** [comment]
> please resolve conflicts

Done

**Comment by rlhui** [comment]
@j-bos , @rck-innovium  please revisit if you'd approve this?  Thanks.

**Review by rck-innovium — APPROVED** [review]

**Comment by rlhui** [comment]
@JaiOCP please close this one. Thanks.

**Pull request closed without merging**